ml_peg/app/utils/plot_helpers.py

```diff
@@ -0,0 +1,439 @@
+            text=labels,
+            customdata=customdata,
+            points="all",
+            jitter=0.05,
+            box_visible=True,
+            meanline_visible=True,
+            fillcolor=color,
+            line_color=color,
+            opacity=0.6,
+            hovertemplate=hovertemplate,
+        )
+    )
+    fig.update_layout(title=title, yaxis_title=yaxis_title, plot_bgcolor="#ffffff")
+    return fig
+
+
+def build_confusion_heatmap(
+    matrix: Sequence[Sequence[float]],
+    *,
+    x_labels: Sequence[str],
+    y_labels: Sequence[str],
+    title: str,
+    colorscale: str = "Blues",
+    hovertemplate: str | None = None,
+) -> go.Figure:
+    """
+    Build a confusion-matrix heatmap with annotated counts.
+
+    Parameters
+    ----------
+    matrix
+        2D array-like of counts.
+    x_labels
+        Labels shown along the x-axis.
+    y_labels
+        Labels shown along the y-axis.
+    title
+        Figure title.
+    colorscale
+        Plotly colourscale applied to the heatmap.
+    hovertemplate
+        Optional template overriding the default hover text.
+
+    Returns
+    -------
+    go.Figure
+        Heatmap with value annotations.
+    """
+    data = np.asarray(matrix, dtype=float)
+    fig = go.Figure()
+    if data.size == 0:
+        fig.update_layout(title=title or "No data available.", plot_bgcolor="#ffffff")
+        return fig
+    template = (
+        hovertemplate or "Pred %{y}<br>Ref %{x}<br>Count: %{z:.0f}<extra></extra>"
+    )
+    fig.add_trace(
+        go.Heatmap(
+            z=data,
+            x=x_labels,
+            y=y_labels,
+            colorscale=colorscale,
+            showscale=True,
+            hovertemplate=template,
+        )
+    )
+    max_val = data.max(initial=0.0)
+    total = data.sum()
+    for y_idx, y_label in enumerate(y_labels):
+        for x_idx, x_label in enumerate(x_labels):
+            cell_val = data[y_idx, x_idx]
+            pct = (cell_val / total * 100) if total else 0.0
+            color = "#ffffff" if max_val and cell_val >= 0.6 * max_val else "#111111"
+            fig.add_annotation(
+                x=x_label,
+                y=y_label,
+                text=f"{cell_val:.0f} ({pct:.1f}%)",
+                showarrow=False,
+                font={"color": color, "size": 14},
+            )
+    fig.update_layout(title=title, plot_bgcolor="#ffffff")
+    return fig
+
+
+def build_classified_parity_scatter(
+    points: Sequence[Mapping[str, float]],
+    *,
+    title: str,
+    xaxis_title: str,
+    yaxis_title: str,
+    class_key: str = "class",
+    ref_key: str = "ref",
+    pred_key: str = "pred",
+    label_key: str = "label",
+    hovertemplate: str,
+    colours: Mapping[str, str] | None = None,
+) -> go.Figure:
+    """
+    Build a scatter plot grouping points by classification labels.
+
+    Parameters
+    ----------
+    points
+        Sequence of metadata dictionaries containing reference/prediction values.
```

Comment:
This is a bit confusing because it makes it sound like data rather than metadata, but I think it's because we save various things (including but not limited to the x/y values) of each point in these dicts? Probably worth a making the connection to "points" in this docstring too, since out of context it's quite confusing